amivapi/auth/sessions.py

```diff
@@ -0,0 +1,204 @@
+        """Allow users to modify only their own sessions."""
+        # item['user'] is Objectid, convert to str
+        return user_id == str(item['user'])
+
+    def create_user_lookup_filter(self, user_id):
+        """Allow users to only see their own sessions."""
+        return {'user': user_id}
+
+
+sessiondomain = {
+    'sessions': {
+        'description': {
+            'general': "A session is used to authenticate a user after he "
+            " provided login data. A POST to /session will return a token you "
+            "can use in an Authorization header: token <yourtoken>",
+            'methods': {
+                'POST': "Login and aquire a login token. Post the fields "
+                "'username' and 'password', the response will contain the "
+                "token. username can be either nethz, mail, or user_id",
+                'GET': "Check token(s)."}
+        },
+
+        'authentication': SessionAuth,
+        'public_methods': ['POST'],
+        'public_item_methods': [],
+        'resource_methods': ['GET', 'POST', 'DELETE'],
+        'item_methods': ['GET', 'DELETE'],
+
+        'schema': {
+            'username': {
+                'type': 'string',
+                'required': True,
+                'nullable': False,
+                'empty': False},
+            'password': {
+                'type': 'string',
+                'required': True,
+                'nullable': False,
+                'empty': False},
+            'user': {'type': 'objectid',
+                     'data_relation': {
+                         'resource': 'users',
+                         'field': '_id',
+                         'embeddable': True,
+                         'cascade_delete': True
+                     },
+                     'readonly': True},
+            'token': {'type': 'string',
+                      'readonly': True}
+        },
+    }
+}
+
+
+# Login Hook
+
+def process_login(items):
+    """Hook to add token on POST to /sessions.
+
+    Attempts login via LDAP if enabled first, then login via database.
+
+    Root login is possible if 'user' is 'root' (instead of nethz or mail).
+    This shortcut is hardcoded.
+
+    TODO (ALEX): make root user shortcut a setting maybe.
+
+    If the login is successful, the fields "username" and "password" are
+    removed and the fields "user" and "token" are added, which will be stored
+    in the db.
+
+    If the login is unsuccessful, abort(401)
+
+    Args:
+        items (list): List of items as passed by EVE to post hooks.
+    """
+    for item in items:  # TODO (Alex): Batch POST doesnt really make sense
+        username = item['username']
+        password = item['password']
+
+        # LDAP
+        if (config.ENABLE_LDAP and
+                ldap_connector.authenticate_user(username, password)):
+            # Success, sync user and get token
+            updated = ldap_connector.sync_one(username)
+            _prepare_token(item, updated['_id'])
```

Comment:
_id is how mongo and eve use the field. Eve sqlalchemy changed this to id. Now it's_id again because eve can handle this without modifications.